src/zod.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  appendErrors,
+  FieldValues,
+  Resolver,
+  ResolverError,
+  ResolverSuccess,
+  transformToNestObject,
+} from 'react-hook-form';
+import { Schema, ZodError, ZodTypeDef } from 'zod';
+import { ParseParams } from 'zod/lib/src/parser';
+import convertArrayToPathName from './utils/convertArrayToPathName';
+
+const parseErrorSchema = (
+  zodError: ZodError,
+  validateAllFieldCriteria: boolean,
+) => {
+  if (zodError.isEmpty) {
+    return {};
+  }
+
+  return zodError.errors.reduce<Record<string, any>>(
+    (previous, { path, message, code: type }) => {
+      const currentPath = convertArrayToPathName(path);
+
+      return {
+        ...previous,
+        ...(path
+          ? previous[currentPath] && validateAllFieldCriteria
+            ? {
+                [currentPath]: appendErrors(
+                  currentPath,
+                  validateAllFieldCriteria,
+                  previous,
+                  type,
+                  message,
+                ),
+              }
+            : {
+                [currentPath]: previous[currentPath] || {
+                  message,
+                  type,
+                  ...(validateAllFieldCriteria
+                    ? {
+                        types: { [type]: message || true },
+                      }
+                    : {}),
+                },
+              }
+          : {}),
+      };
+    },
+    {},
+  );
+};
+
+export const zodResolver = <
+  TFieldValues extends FieldValues,
+  TZodTypeDef extends ZodTypeDef
+>(
+  schema: Schema<TFieldValues, TZodTypeDef>,
```

Comment:
I recommend writing generic functions that accept Zod schemas like so:

```ts
const zodResolver = <T extends z.ZodSchema<any,any>>(
  valueSchema: T,
): Resolver<Infer<T>> => { // you'll have to import `Infer` from 'zod'
  // do stuff
}
```

I talk about the reasons a bit here: https://github.com/vriad/zod/issues/153#issuecomment-692903583

Reply:
Thanks for your recommandations !
I've used `TypeOf` from zod instead of `Infer` to avoid this import:  `import { Infer } from 'zod/lib/src/types/base';`
I'm wrong ?

Reply:
Oops typo: it should be `infer` (lower case)

And they're exactly the same 😅   I added `infer` as an alias because I thought it was a little more intuitive.

![Screen Shot 2020-09-16 at 12 02 46 PM](https://user-images.githubusercontent.com/3084745/93380773-899ca400-f814-11ea-8356-b57d72d61e9d.png)

Reply:
Thanks ! updated ! :)